src/picongpu/include/particles/ionization/ionization.hpp

```diff
@@ -0,0 +1,397 @@
+    if (linearThreadIdx == 0)
+    {
+        ionFrame = &(ionBox.getLastFrame(block, isValid));
+        maxParticlesInFrame = PMacc::math::CT::volume<SuperCellSize>::type::value;
+    }
+
+    __syncthreads();
+    if (!isValid)
+        return; //end kernel if we have no frames
+
+    /* caching of E and B fields */
+    PMACC_AUTO(cachedB, CachedBox::create < 0, typename BBox::ValueType > (BlockDescription_()));
+    PMACC_AUTO(cachedE, CachedBox::create < 1, typename EBox::ValueType > (BlockDescription_()));
+
+    __syncthreads();
+    
+    nvidia::functors::Assign assign;
+    
+    PMACC_AUTO(fieldBBlock, fieldB.shift(blockCell));
+    ThreadCollective<BlockDescription_> collectiv(linearThreadIdx);
+    collectiv(
+              assign,
+              cachedB,
+              fieldBBlock
+              );
+
+    PMACC_AUTO(fieldEBlock, fieldE.shift(blockCell));
+    collectiv(
+              assign,
+              cachedE,
+              fieldEBlock
+              );
+    __syncthreads();
+
+    /* Declare counter in shared memory that will later tell the current fill level or 
+     * occupation of the newly created target electron frames. */
+    __shared__ int newFrameFillLvl;
+    
+    __syncthreads(); /*wait that all shared memory is initialized*/
+    
+    /* Declare local variable oldFrameFillLvl for each thread */
+    int oldFrameFillLvl;
+    
+    /* Initialize local (register) counter for each thread 
+     * - describes how many new electrons should be created */
+    unsigned int newElectrons = 0;
+    
+    /* Declare local electron ID
+     * - describes at which position in the new frame the new electron is to be created */
+    int electronId;
+    
+    /* Master initializes the frame fill level with 0 */
+    if (linearThreadIdx == 0)
+    {
+        newFrameFillLvl = 0;        
+        electronFrame = NULL;
+    }
+    __syncthreads();
+    
+    /* move over particle frames and call frameIonizer 
+     * frames are worked on in backwards order to avoid asking if there is another frame
+     * --> performance 
+     * Because all frames are completely filled except the last and apart from that last frame 
+     * one wants to make sure that all threads are working and every frame is worked on. */
+    while (isValid)
+    {
+        /* casting uint8_t multiMask to boolean */
+        bool isParticle = (*ionFrame)[linearThreadIdx][multiMask_];
+        __syncthreads();;
+
+        /* < IONIZATION and change of charge states > 
+         * if the threads contain particles, the frameIonizer can ionize them 
+         * if they are non-particles their inner ionization counter remains at 0 */
+        if (isParticle)
+        {
+        /* ionization based on ionization model - this actually increases charge states*/
+        frameIonizer(*ionFrame, linearThreadIdx, cachedB, cachedE, newElectrons);
+        }
+        __syncthreads();
+        /* always true while-loop over all source frames */
+        while (true)
```

Comment:
The functor business has been stashed to a different branch because I couldn't make it work so quickly. It can become included in a later pull request.